File: backend/app/models/project.py
from sqlalchemy import Column, String, Text, Integer, Boolean, DateTime, ForeignKey, Enum as SQLEnum
from sqlalchemy.dialects.postgresql import UUID, JSON
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
from app.core.database import Base
from app.models.enums import LocationType
import uuid
import enum


class ProjectCategory(str, enum.Enum):
    """プロジェクトカテゴリ"""
    ASOBI = "asobi"  # あそびプロジェクト
    ASOTO = "asoto"  # あそとプロジェクト


class ProjectStatus(str, enum.Enum):
    """プロジェクトステータス"""
    RECRUITING = "recruiting"  # 募集中
    ACTIVE = "active"  # 進行中
    COMPLETED = "completed"  # 完了
    ARCHIVED = "archived"  # アーカイブ


class ProjectVisibility(str, enum.Enum):
    """プロジェクト公開設定"""
    PUBLIC = "public"  # 公開
    MEMBERS_ONLY = "members_only"  # メンバーのみ


class Project(Base):
    __tablename__ = "projects"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    owner_id = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False)

    # 基本情報
    title = Column(String(255), nullable=False)
    description = Column(Text)
    category = Column(SQLEnum(ProjectCategory), nullable=False)
    status = Column(SQLEnum(ProjectStatus), default=ProjectStatus.RECRUITING)

    # 期間・活動
    start_date = Column(DateTime(timezone=True), nullable=False)
    end_date = Column(DateTime(timezone=True))
    frequency = Column(String(255))  # "週末月2回程度"

    # 場所
    location_type = Column(SQLEnum(LocationType), nullable=False)
    location_detail = Column(String(500))

    # メンバー募集
    is_recruiting = Column(Boolean, default=False)
    max_members = Column(Integer)
    required_skills = Column(JSON, default=list)  # ["農業知識", "写真撮影"]

    # メタ情報
    tags = Column(JSON, default=list)  # ["農業", "地域", "暮らし"]
    visibility = Column(SQLEnum(ProjectVisibility), default=ProjectVisibility.PUBLIC)

    # タイムスタンプ
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())

    # リレーション
    owner = relationship("User", back_populates="owned_projects")
    members = relationship("ProjectMember", back_populates="project", cascade="all, delete-orphan")
    tasks = relationship("ProjectTask", back_populates="project", cascade="all, delete-orphan")

File: frontend/src/types/project.ts
export type ProjectCategory = 'asobi' | 'asoto';
export type ProjectStatus = 'planning' | 'active' | 'paused' | 'completed' | 'cancelled';
export type ProjectVisibility = 'public' | 'private' | 'members_only';
export type MemberRole = 'owner' | 'admin' | 'member' | 'viewer';
export type MemberStatus = 'pending' | 'active' | 'inactive';
export type TaskStatus = 'todo' | 'in_progress' | 'done';

export interface Project {
  id: string;
  owner_id: string;
  title: string;
  description?: string | null;
  category: ProjectCategory;
  status: ProjectStatus;
  start_date: string;
  end_date?: string | null;
  frequency?: string | null;
  location_type: 'online' | 'offline' | 'hybrid';
  location_detail?: string | null;
  is_recruiting: boolean;
  max_members?: number | null;
  required_skills: string[];
  tags: string[];
  visibility: ProjectVisibility;
  created_at: string;
  updated_at: string;
}

export interface CreateProjectInput {
  title: string;
  description?: string;
  category: ProjectCategory;
  start_date: string;
  end_date?: string;
  frequency?: string;
  location_type: 'online' | 'offline' | 'hybrid';
  location_detail?: string;
  is_recruiting?: boolean;
  max_members?: number;
  required_skills?: string[];
  tags?: string[];
  visibility?: ProjectVisibility;
}

export interface UpdateProjectInput {
  title?: string;
  description?: string;
  category?: ProjectCategory;
  status?: ProjectStatus;
  start_date?: string;
  end_date?: string;
  frequency?: string;
  location_type?: 'online' | 'offline' | 'hybrid';
  location_detail?: string;
  is_recruiting?: boolean;
  max_members?: number;
  required_skills?: string[];
  tags?: string[];
  visibility?: ProjectVisibility;
}
